### PR TITLE
feat(eventLoop): the visibility of the shutdown flag for the event loop and optimizations for graceful shutdown

### PR DESCRIPTION
**Pull request by CLFutureX**
The visibility of the shutdown flag for the event loop and optimizations for graceful shutdown

### CODE REVIEW

File: s3stream/src/main/java/com/automq/stream/utils/threads/EventLoop.java [hunk 1]
```diff
@@ -79,6 +89,9 @@ public void execute(Runnable task) {
 
     public CompletableFuture<Void> shutdownGracefully() {
         shutdown = true;
+        if (!shutdownCf.isDone()) {
```
Comment (superhx) [inline s3stream/src/main/java/com/automq/stream/utils/threads/EventLoop.java:92]: To ensure concurrent safety, the operations of modifying `shutdown` and `tasks` need to be safeguarded by a lock. Similarly, other methods that use these two variables need to be modified.

Comment (CLFutureX) [reply s3stream/src/main/java/com/automq/stream/utils/threads/EventLoop.java]: > To ensure concurrent safety, the operations of modifying `shutdown` and `tasks` need to be safeguarded by a lock. Similarly, other methods that use these two variables need to be modified.

Yes, perhaps we can modify the shutdown from a constant to an AtomicBoolean, utilizing CAS (Compare-And-Swap) to ensure concurrent safety.

Comment (CLFutureX) [reply s3stream/src/main/java/com/automq/stream/utils/threads/EventLoop.java]: > To ensure concurrent safety, the operations of modifying `shutdown` and `tasks` need to be safeguarded by a lock. Similarly, other methods that use these two variables need to be modified.

now, the tasks variable has also been modified.

**Comment by CLFutureX** [comment]
@superhx   Perhaps  it should only adjust the visibility of the shutdown variable, without introducing a wakeup mechanism, because the worst-case scenario would only involve a wait of 100ms, which is generally acceptable for a graceful shutdown, What do you think?

File: s3stream/src/main/java/com/automq/stream/utils/threads/EventLoop.java [hunk 2]
```diff
@@ -17,14 +17,21 @@
...
 import java.util.concurrent.LinkedBlockingQueue;
 import java.util.concurrent.TimeUnit;
+
 import org.slf4j.Logger;
 
 public class EventLoop extends Thread implements Executor {
     private final Logger logger;
     private BlockingQueue<Runnable> tasks;
-    private boolean shutdown = false;
+    private volatile boolean shutdown;
```
Comment (superhx) [inline s3stream/src/main/java/com/automq/stream/utils/threads/EventLoop.java:26]: Did not see when shutdown = true was set

**Comment by superhx** [comment]
> @superhx Perhaps it should only adjust the visibility of the shutdown variable, without introducing a wakeup mechanism, because the worst-case scenario would only involve a wait of 100ms, which is generally acceptable for a graceful shutdown, What do you think?

If it's not completed, I think it's ok to add the shutdown awake.

**Comment by CLFutureX** [comment]
@superhx  hey, please take a look

**Review by superhx — APPROVED** [review]

**Pull request merged**